src/solvers/refinement/string_constraint_generator_float.cpp

```diff
@@ -0,0 +1,454 @@
+/*******************************************************************\
+
+Module: Generates string constraints for functions generating strings
+        from other types, in particular int, long, float, double, char, bool
+
+Author: Romain Brenguier, romain.brenguier@diffblue.com
+
+\*******************************************************************/
+
+/// \file
+/// Generates string constraints for functions generating strings from other
+/// types, in particular int, long, float, double, char, bool
+
+#include <solvers/floatbv/float_bv.h>
+
+#include "string_constraint_generator.h"
+
+
+/// Gets the unbiased exponent in a floating-point bit-vector
+///
+/// TODO: factorize with float_bv.cpp float_utils.h
+/// \param src: a floating point expression
+/// \param spec: specification for floating points
+/// \return A new 32 bit integer expression representing the exponent.
+///   Note that 32 bits are sufficient for the exponent even
+///   in octuple precision.
+exprt get_exponent(
+  const exprt &src, const ieee_float_spect &spec)
+{
+  exprt exp_bits=extractbits_exprt(
+    src, spec.f+spec.e-1, spec.f, unsignedbv_typet(spec.e));
+
+  // Exponent is in biased from (numbers from -128 to 127 are encoded with
+  // integer from 0 to 255) we have to remove the bias.
+  return minus_exprt(
+    typecast_exprt(exp_bits, signedbv_typet(32)),
+    from_integer(spec.bias(), signedbv_typet(32)));
+}
+
+/// Gets the magnitude without hidden bit
+/// \param src: a floating point expression
+/// \param spec: specification for floating points
+/// \return An unsigned value representing the magnitude.
+exprt get_magnitude(const exprt &src, const ieee_float_spect &spec)
```

Comment:
Double-check terminology -- this is getting the mantissa bits, whereas "magnitude" suggests the exponent to me more than the significand.